osu.Game/Online/API/APIDownloadRequest.cs

```diff
@@ -28,13 +29,19 @@ protected override WebRequest CreateWebRequest()
 
         private void request_Progress(long current, long total) => API.Schedule(() => Progressed?.Invoke(current, total));
 
-        protected APIDownloadRequest()
+        internal void TriggerSuccess(string filename)
         {
-            base.Success += onSuccess;
+            if (this.filename != null)
+                throw new InvalidOperationException("Attempted to trigger success more than once");
+
+            this.filename = filename;
+
+            TriggerSuccess();
         }
```

Comment:
why does this exist in the base request, instead of near the test where this actually makes sense?

Reply:
API requests generally expose an internal `TriggerSuccess` method for testing purposes (even `APIRequest<T>` where it allows to provide arbitrary content like in here as the result of the successful request), so why bother doing it in the test request if it should've been in the base request class itself anyways.

Reply:
disagree. the filename is an internal implementation detail of the request. fundamental difference between controlling the core content of the response for a test, and being able to control a temporary filename in a download request for a test.

Reply:
fair point